dwa_local_planner/src/dwa_planner.cpp

```diff
@@ -252,8 +259,87 @@ namespace dwa_local_planner {
     return false;
   }
 
+  uint32_t DWAPlanner::cropPlanToReachableGoal(
+    const geometry_msgs::PoseStamped& global_pose,
+    const std::vector<geometry_msgs::Point>& footprint_spec,
+    std::vector<geometry_msgs::PoseStamped>& plan) {
+    // search the path for a reachable goal
+    const double dist_to_goal = mbf_utility::distance(global_pose, plan.back());
+
+    // find furthest point along the path of which the corresponding cell is in free space
+    const costmap_2d::Costmap2D* const costmap = planner_util_->getCostmap();
+    int last_point_on_the_map = -1;
+    int last_point_in_free_space = -1;
+    for (unsigned int i = 0; i < plan.size(); ++i) {
+      double g_x = plan[i].pose.position.x;
+      double g_y = plan[i].pose.position.y;
+      unsigned int map_x, map_y;
+      if (!costmap->worldToMap(g_x, g_y, map_x, map_y)) {
+        // give up once a point within the map had been found
+        if (last_point_on_the_map >= 0) {
+          break;
+        }
+        continue;
+      }
 
-  void DWAPlanner::updatePlanAndLocalCosts(
+      last_point_on_the_map = i;
+      if (costmap->getCost(map_x, map_y) < costmap_2d::INSCRIBED_INFLATED_OBSTACLE) {
+        last_point_in_free_space = i;
+      }
+    }
+
+    // handle general errors
+    if (last_point_on_the_map == -1) {
+      ROS_ERROR("None of the points of the global plan were in the local costmap, global plan points too far from robot");
+      return mbf_msgs::ExePathResult::OUT_OF_MAP;
+    }
+
+    if (last_point_in_free_space == -1) {
+      ROS_ERROR("None of the points of the global plan were in free space");
+      return mbf_msgs::ExePathResult::BLOCKED_PATH;
+    }
+
+    const auto dist_to_last_free_point = mbf_utility::distance(global_pose, plan[last_point_in_free_space]);
+    const auto limits = planner_util_->getCurrentLimits();
+    if (dist_to_last_free_point > limits.goal_obstacle_approach_distance) {
```

Comment:
Q: I am a bit confused here. So the goal_obstacle_approach_dist is how much we navigate on the path to accept it as success? Because you compute the distance from robot to last free space, not distance from goal to last free space.
In this case, I think the name of the variable is a bit confusing, it seems goal<->obstacle distance.
Besides that, shouldn't we check footprint clearance also? Maybe change the first loop to check footprint clearance right away, instead of having another loop to check that.

Reply:
Does the description of the "goal_obstacle_approach_distance"` answer this question? `"In case the goal is blocked, move towards it until the obstacle blocking the goal is within this distance"`

Basically, if far enough from the obstacle, don't even care about footprint checks, and just crop the plan such that the last point of the plan is in a cell in free space.

Then once getting to the proximity, do the actual footprint checks, resulting in the behavior of e.g. the third video, where the robot goes towards the goal, and gives up once it is within one meter. (The reasoning for this behavior is when the goal is still far away, it could be blocked according to the costmap due to some obstacles marked earlier; to ensure whether the goal is actually still blocked, the robot needs to go into the proximity of the goal)

Reply:
hum, ok, i suggest to add that comment here then (the reasoning for this behavior is when the goal is still far away, it could be blocked according to the costmap due to some obstacles marked earlier; to ensure whether the goal is actually still blocked, the robot needs to go into the proximity of the goal), because we may forget why.

Reply:
Ok, added comments in https://github.com/rapyuta-robotics/ros-navigation/pull/74/commits/cf50d382e68337a7093718d8bb16ccef67703dc7